lib/catalog/api.go

```diff
@@ -89,31 +93,52 @@ type DownloadResult struct {
 	Path string
 }
 
-// Downloads downloads the specified application and returns its path.
+// Close removes the downloaded tarball.
+//
+// Implements io.Closer.
+func (r *DownloadResult) Close() error {
+	return os.RemoveAll(filepath.Dir(r.Path))
+}
+
+// Download downloads the specified application and returns its path.
 func Download(req DownloadRequest) (*DownloadResult, error) {
 	log.Debugf("%#v", req)
 	localCluster, err := localenv.LocalCluster()
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 	var catalog Catalog
-	switch req.Locator.Repository {
+	switch req.Application.Repository {
 	case "", localCluster.Domain: // local cluster app is requested
 		catalog, err = NewLocal()
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
 	default: // app from remote catalog (Ops Center) is requested
-		catalog, err = NewRemoteFor(req.Locator.Repository)
+		catalog, err = NewRemoteFor(req.Application.Repository)
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
 	}
-	path, err := catalog.Download(req.Locator.Name, req.Locator.Version)
+	reader, err := catalog.Download(req.Application.Name, req.Application.Version)
```

Comment:
Not sure how it'd be more efficient - it would just move creation of the temp. file/directory to the caller side (so it can be passed to download request) which would also result in more code (cause this method is called twice). There's no intermediate copy - the caller uses the tarball at the path this method downloaded it at.